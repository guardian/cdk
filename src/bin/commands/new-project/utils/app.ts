import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
