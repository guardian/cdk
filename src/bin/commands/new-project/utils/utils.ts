import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
