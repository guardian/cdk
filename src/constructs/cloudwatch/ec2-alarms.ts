import { ComparisonOperator, MathExpression, Statistic, TreatMissingData } from "@aws-cdk/aws-cloudwatch";
import { HttpCodeElb, HttpCodeTarget } from "@aws-cdk/aws-elasticloadbalancingv2";
import { Duration } from "@aws-cdk/core";
import { AppIdentity } from "../core/identity";
import { GuAlarm } from "./alarm";
import type { GuStack } from "../core";
import type { GuApplicationLoadBalancer, GuApplicationTargetGroup } from "../loadbalancing";
import type { GuAlarmProps } from "./alarm";

export interface Http5xxAlarmProps
  extends Omit<
    GuAlarmProps,
    "snsTopicName" | "evaluationPeriods" | "metric" | "period" | "threshold" | "treatMissingData"
  > {
  tolerated5xxPercentage: number;
  numberOfMinutesAboveThresholdBeforeAlarm?: number;
}

interface Gu5xxPercentageAlarmProps extends Pick<GuAlarmProps, "snsTopicName">, Http5xxAlarmProps, AppIdentity {
  loadBalancer: GuApplicationLoadBalancer;
}

interface GuUnhealthyInstancesAlarmProps extends Pick<GuAlarmProps, "snsTopicName">, AppIdentity {
  targetGroup: GuApplicationTargetGroup;
}

/**
 * Creates an alarm which is triggered whenever the percentage of requests with a 5xx response code exceeds
 * the specified threshold.
 */
export class Gu5xxPercentageAlarm extends GuAlarm {
  constructor(scope: GuStack, props: Gu5xxPercentageAlarmProps) {
    const mathExpression = new MathExpression({
      expression: "100*(m1+m2)/m3",
      usingMetrics: {
        m1: props.loadBalancer.metricHttpCodeElb(HttpCodeElb.ELB_5XX_COUNT),
        m2: props.loadBalancer.metricHttpCodeTarget(HttpCodeTarget.TARGET_5XX_COUNT),
        m3: props.loadBalancer.metricRequestCount(),
      },
      label: `% of 5XX responses served for ${props.app} (load balancer and instances combined)`,
      period: Duration.minutes(1),
    });
    const defaultAlarmName = `High 5XX error % from ${props.app} in ${scope.stage}`;
    const defaultDescription = `${props.app} exceeded ${props.tolerated5xxPercentage}% error rate`;
    const alarmProps = {
      ...props,
      metric: mathExpression,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      threshold: props.tolerated5xxPercentage,
      comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
      alarmName: props.alarmName ?? defaultAlarmName,
      alarmDescription: props.alarmDescription ?? defaultDescription,
      evaluationPeriods: props.numberOfMinutesAboveThresholdBeforeAlarm ?? 1,
    };
    super(scope, AppIdentity.suffixText(props, "High5xxPercentageAlarm"), alarmProps);
  }
}

/**
 * Creates an alarm which is triggered whenever there have been several healthcheck failures within a single hour.
 */
export class GuUnhealthyInstancesAlarm extends GuAlarm {
  constructor(scope: GuStack, props: GuUnhealthyInstancesAlarmProps) {
    const alarmName = `Unhealthy instances for ${props.app} in ${scope.stage}`;
    const alarmDescription = `${props.app}'s instances have failed healthchecks several times over the last hour.
      This typically results in the AutoScaling Group cycling instances and can lead to problems with deployment,
      scaling or handling traffic spikes.

      Check ${props.app}'s application logs or ssh onto an unhealthy instance in order to debug these problems.`;
    const alarmProps = {
      ...props,
      alarmName: alarmName,
      alarmDescription: alarmDescription,
      metric: props.targetGroup.metricUnhealthyHostCount(),
      treatMissingData: TreatMissingData.NOT_BREACHING,
      statistic: Statistic.MAXIMUM,
      threshold: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      period: Duration.minutes(5),
      datapointsToAlarm: 6,
      evaluationPeriods: 12,
    };
    super(scope, AppIdentity.suffixText(props, "UnhealthyInstancesAlarm"), alarmProps);
  }
}
