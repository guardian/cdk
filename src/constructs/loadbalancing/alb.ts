import type {
  ApplicationListenerProps,
  ApplicationLoadBalancerProps,
  ApplicationTargetGroupProps,
  CfnListener,
  CfnLoadBalancer,
  CfnTargetGroup,
} from "@aws-cdk/aws-elasticloadbalancingv2";
import {
  ApplicationListener,
  ApplicationLoadBalancer,
  ApplicationProtocol,
  ApplicationTargetGroup,
  ListenerAction,
  Protocol,
} from "@aws-cdk/aws-elasticloadbalancingv2";
import { Duration } from "@aws-cdk/core";
import { RegexPattern } from "../../constants";
import type { GuStack } from "../core";
import { GuCertificateArnParameter } from "../core";
import type { AppIdentity } from "../core/identity";

interface GuApplicationLoadBalancerProps extends ApplicationLoadBalancerProps {
  overrideId?: boolean;
}

export class GuApplicationLoadBalancer extends ApplicationLoadBalancer {
  constructor(scope: GuStack, id: string, props: GuApplicationLoadBalancerProps) {
    super(scope, id, { deletionProtection: true, ...props });

    const cfnLb = this.node.defaultChild as CfnLoadBalancer;

    if (props.overrideId || (scope.migratedFromCloudFormation && props.overrideId !== false))
      cfnLb.overrideLogicalId(id);

    cfnLb.addPropertyDeletionOverride("Type");
  }
}

export interface GuApplicationTargetGroupProps extends ApplicationTargetGroupProps {
  overrideId?: boolean;
}

export class GuApplicationTargetGroup extends ApplicationTargetGroup {
  static DefaultHealthCheck = {
    path: "/healthcheck",
    protocol: Protocol.HTTP,
    healthyThresholdCount: 2,
    unhealthyThresholdCount: 5,
    interval: Duration.seconds(30),
    timeout: Duration.seconds(10),
  };

  constructor(scope: GuStack, id: string, props: GuApplicationTargetGroupProps) {
    const mergedProps = {
      ...props,
      healthCheck: { ...GuApplicationTargetGroup.DefaultHealthCheck, ...props.healthCheck },
    };

    super(scope, id, mergedProps);

    if (mergedProps.overrideId || (scope.migratedFromCloudFormation && mergedProps.overrideId !== false))
      (this.node.defaultChild as CfnTargetGroup).overrideLogicalId(id);
  }
}

export interface GuApplicationListenerProps extends ApplicationListenerProps {
  overrideId?: boolean;
}

export class GuApplicationListener extends ApplicationListener {
  constructor(scope: GuStack, id: string, props: GuApplicationListenerProps) {
    super(scope, id, { port: 443, protocol: ApplicationProtocol.HTTPS, ...props });

    if (props.overrideId || (scope.migratedFromCloudFormation && props.overrideId !== false))
      (this.node.defaultChild as CfnListener).overrideLogicalId(id);
  }
}

export interface GuHttpsApplicationListenerProps
  extends Omit<GuApplicationListenerProps, "defaultAction" | "certificates">,
    AppIdentity {
  targetGroup: GuApplicationTargetGroup;
  certificate?: string;
}

export class GuHttpsApplicationListener extends ApplicationListener {
  constructor(scope: GuStack, id: string, props: GuHttpsApplicationListenerProps) {
    if (props.certificate) {
      const isValid = new RegExp(RegexPattern.ACM_ARN).test(props.certificate);
      if (!isValid) {
        throw new Error(`${props.certificate} is not a valid ACM ARN`);
      }
    }

    const mergedProps: GuApplicationListenerProps = {
      port: 443,
      protocol: ApplicationProtocol.HTTPS,
      ...props,
      certificates: [
        {
          certificateArn: props.certificate ?? new GuCertificateArnParameter(scope, props).valueAsString,
        },
      ],
      defaultAction: ListenerAction.forward([props.targetGroup]),
    };

    super(scope, id, mergedProps);
  }
}
