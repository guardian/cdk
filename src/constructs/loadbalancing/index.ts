export * from "./alb";
export * from "./elb";
