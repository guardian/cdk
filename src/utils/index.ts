export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
