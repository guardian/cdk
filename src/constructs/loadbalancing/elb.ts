import type {
  CfnLoadBalancer,
  HealthCheck,
  LoadBalancerListener,
  LoadBalancerProps,
} from "@aws-cdk/aws-elasticloadbalancing";
import { LoadBalancer, LoadBalancingProtocol } from "@aws-cdk/aws-elasticloadbalancing";
import { Duration } from "@aws-cdk/core";
import { GuStatefulMigratableConstruct } from "../../utils/mixin";
import type { GuStack } from "../core";
import { GuArnParameter } from "../core";
import { AppIdentity } from "../core/identity";
import type { GuMigratingResource } from "../core/migrating";

interface GuClassicLoadBalancerProps extends Omit<LoadBalancerProps, "healthCheck">, GuMigratingResource, AppIdentity {
  propertiesToOverride?: Record<string, unknown>;
  healthCheck?: Partial<HealthCheck>;
}

/**
 * **IMPORTANT**: This construct should **only** be used if you are migrating an existing stack and you need to retain the load balancer.
 * Please use [[`GuApplicationLoadBalancer`]] instead of [[`GuClassicLoadBalancer`]] wherever possible.
 *
 * In order to inherit an existing load balancer, the `migratedFromCloudFormation` prop on your stack must be set to `true`.
 * You must also pass the logical id from your CloudFormation template to this construct via the `existingLogicalId` prop.
 *
 * By default, load balancers created via this construct will perform a healthcheck against `/healthcheck` on port 9000. All healthcheck
 * defaults can be overridden via the `healthcheck` prop.
 *
 * For example, to use `/test` for the healthcheck path use:
 *
 * ```typescript
 * new GuClassicLoadBalancer(stack, "ClassicLoadBalancer", {
 *     // other props
 *     healthCheck: {
 *       path: "/test",
 *     },
 *   });
 * ```
 * If you are running an application which only accepts traffic over HTTPs, consider using [[`GuHttpsClassicLoadBalancer`]]
 * to reduce the amount of boilerplate needed when configuring your load balancer.
 */
export class GuClassicLoadBalancer extends GuStatefulMigratableConstruct(LoadBalancer) {
  static DefaultHealthCheck = {
    port: 9000,
    path: "/healthcheck",
    protocol: LoadBalancingProtocol.HTTP,
    healthyThreshold: 2,
    unhealthyThreshold: 5,
    interval: Duration.seconds(30),
    timeout: Duration.seconds(10),
  };

  constructor(scope: GuStack, id: string, props: GuClassicLoadBalancerProps) {
    const mergedProps = {
      ...props,
      healthCheck: { ...GuClassicLoadBalancer.DefaultHealthCheck, ...props.healthCheck },
    };

    super(scope, AppIdentity.suffixText({ app: props.app }, id), mergedProps);
    AppIdentity.taggedConstruct({ app: props.app }, this);

    const cfnLb = this.node.defaultChild as CfnLoadBalancer;

    mergedProps.propertiesToOverride &&
      Object.entries(mergedProps.propertiesToOverride).forEach(([key, value]) => cfnLb.addPropertyOverride(key, value));
  }
}

interface GuHttpsClassicLoadBalancerProps extends Omit<GuClassicLoadBalancerProps, "listeners"> {
  listener?: Partial<LoadBalancerListener>;
}

/**
 * **IMPORTANT**: This construct should **only** be used if you are migrating an existing stack and you need to retain the load balancer.
 * Please use [[`GuHttpsApplicationListener`]] instead of [[`GuHttpsClassicLoadBalancer`]] wherever possible.
 *
 * This construct creates a classic load balancer which accepts HTTPS traffic. It communicates with EC2 instances over port 9000
 * by default. This default can be overridden via the listener prop, for example:
 * ```typescript
 * new GuHttpsClassicLoadBalancer(stack, "HttpsClassicLoadBalancer", {
 *     // other props
 *     listener: {
 *       internalPort: 3000,
 *     },
 *   });
 * ```
 *
 * You can pass a certificate id to this construct via the listener interface, for example:
 * ```typescript
 * new GuHttpsClassicLoadBalancer(stack, "HttpsClassicLoadBalancer", {
 *     // other props
 *     listener: {
 *       sslCertificateId: "certificateId",
 *     },
 *   });
 * ```
 * If certificate id is omitted the library will create a Parameter which allows you to pass in a certificate ARN.
 *
 * For more details on migrating an existing load balancer and general load balancer defaults, see [[`GuClassicLoadBalancer`]].
 */
export class GuHttpsClassicLoadBalancer extends GuClassicLoadBalancer {
  static DefaultListener: LoadBalancerListener = {
    internalPort: 9000,
    externalPort: 443,
    internalProtocol: LoadBalancingProtocol.HTTP,
    externalProtocol: LoadBalancingProtocol.HTTPS,
  };

  constructor(scope: GuStack, id: string, props: GuHttpsClassicLoadBalancerProps) {
    const listenerProps = { ...GuHttpsClassicLoadBalancer.DefaultListener, ...props.listener };

    if (!listenerProps.sslCertificateId) {
      const certificateId = new GuArnParameter(scope, "CertificateARN", {
        description: "Certificate ARN for ELB",
      });
      listenerProps.sslCertificateId = certificateId.valueAsString;
    }

    const mergedProps = {
      ...props,
      listeners: [listenerProps],
    };

    super(scope, id, mergedProps);
  }
}
