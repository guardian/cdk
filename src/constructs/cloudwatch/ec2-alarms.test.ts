import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import { ApplicationListener, ApplicationProtocol } from "aws-cdk-lib/aws-elasticloadbalancingv2";
import { simpleGuStackForTesting } from "../../utils/test";
import type { AppIdentity } from "../core";
import { GuApplicationLoadBalancer, GuApplicationTargetGroup } from "../loadbalancing";
import { GuAlb5xxPercentageAlarm, GuUnhealthyInstancesAlarm } from "./ec2-alarms";

const vpc = Vpc.fromVpcAttributes(new Stack(), "VPC", {
  vpcId: "test",
  availabilityZones: [""],
  publicSubnetIds: [""],
});

const app: AppIdentity = {
  app: "testing",
};

describe("The GuAlb5xxPercentageAlarm construct", () => {
  it("should create the correct alarm resource with minimal config", () => {
    const stack = simpleGuStackForTesting();
    const alb = new GuApplicationLoadBalancer(stack, "ApplicationLoadBalancer", { ...app, vpc });
    const props = {
      tolerated5xxPercentage: 1,
      snsTopicName: "test-topic",
    };
    new GuAlb5xxPercentageAlarm(stack, { ...app, loadBalancer: alb, ...props });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should use a custom description if one is provided", () => {
    const stack = simpleGuStackForTesting();
    const alb = new GuApplicationLoadBalancer(stack, "ApplicationLoadBalancer", { ...app, vpc });
    const props = {
      alarmDescription: "test-custom-alarm-description",
      tolerated5xxPercentage: 1,
      snsTopicName: "test-topic",
    };
    new GuAlb5xxPercentageAlarm(stack, { ...app, loadBalancer: alb, ...props });
    Template.fromStack(stack).hasResourceProperties("AWS::CloudWatch::Alarm", {
      AlarmDescription: "test-custom-alarm-description",
    });
  });

  it("should use a custom alarm name if one is provided", () => {
    const stack = simpleGuStackForTesting();
    const alb = new GuApplicationLoadBalancer(stack, "ApplicationLoadBalancer", { ...app, vpc });
    const props = {
      alarmName: "test-custom-alarm-name",
      tolerated5xxPercentage: 1,
      snsTopicName: "test-topic",
    };
    new GuAlb5xxPercentageAlarm(stack, { ...app, loadBalancer: alb, ...props });
    Template.fromStack(stack).hasResourceProperties("AWS::CloudWatch::Alarm", {
      AlarmName: "test-custom-alarm-name",
    });
  });

  it("should adjust the number of evaluation periods if a custom value is provided", () => {
    const stack = simpleGuStackForTesting();
    const alb = new GuApplicationLoadBalancer(stack, "ApplicationLoadBalancer", { ...app, vpc });
    const props = {
      tolerated5xxPercentage: 1,
      numberOfMinutesAboveThresholdBeforeAlarm: 3,
      snsTopicName: "test-topic",
    };
    new GuAlb5xxPercentageAlarm(stack, { ...app, loadBalancer: alb, ...props });
    Template.fromStack(stack).hasResourceProperties("AWS::CloudWatch::Alarm", {
      EvaluationPeriods: 3,
    });
  });
});

describe("The GuUnhealthyInstancesAlarm construct", () => {
  it("should create the correct alarm resource with minimal config", () => {
    const stack = simpleGuStackForTesting();
    const alb = new GuApplicationLoadBalancer(stack, "ApplicationLoadBalancer", { ...app, vpc });
    const targetGroup = new GuApplicationTargetGroup(stack, "ApplicationTargetGroup", { ...app, vpc });
    new ApplicationListener(stack, "ApplicationListener", {
      protocol: ApplicationProtocol.HTTP,
      ...app,
      loadBalancer: alb,
      defaultTargetGroups: [targetGroup],
    });
    new GuUnhealthyInstancesAlarm(stack, { ...app, targetGroup: targetGroup, snsTopicName: "test-topic" });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });
});
