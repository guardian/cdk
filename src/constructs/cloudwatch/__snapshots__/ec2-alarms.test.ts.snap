// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Gu5xxPercentageAlarm construct should create the correct alarm resource with minimal config 1`] = `
Object {
  "Mappings": Object {
    "testing": Object {
      "CODE": Object {
        "alarmActionsEnabled": false,
      },
      "PROD": Object {
        "alarmActionsEnabled": true,
      },
    },
  },
  "Outputs": Object {
    "ApplicationLoadBalancerTestingDnsName": Object {
      "Description": "DNS entry for ApplicationLoadBalancerTesting",
      "Value": Object {
        "Fn::GetAtt": Array [
          "ApplicationLoadBalancerTesting172A253B",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApplicationLoadBalancerTesting172A253B": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internal",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ApplicationLoadBalancerTestingSecurityGroup883A01A4",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          "",
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "ApplicationLoadBalancerTestingSecurityGroup883A01A4": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB TestApplicationLoadBalancerTesting8F9EA5A8",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": "test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "High5xxPercentageAlarmTesting9E960B0F": Object {
      "Properties": Object {
        "ActionsEnabled": Object {
          "Fn::FindInMap": Array [
            "testing",
            Object {
              "Ref": "Stage",
            },
            "alarmActionsEnabled",
          ],
        },
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":test-topic",
              ],
            ],
          },
        ],
        "AlarmDescription": "testing exceeded 1% error rate",
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "High 5XX error % from testing in ",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": Array [
          Object {
            "Expression": "100*(m1+m2)/m3",
            "Id": "expr_1",
            "Label": "% of 5XX responses served for testing (load balancer and instances combined)",
          },
          Object {
            "Id": "m1",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "LoadBalancer",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "ApplicationLoadBalancerTesting172A253B",
                        "LoadBalancerFullName",
                      ],
                    },
                  },
                ],
                "MetricName": "HTTPCode_ELB_5XX_Count",
                "Namespace": "AWS/ApplicationELB",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m2",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "LoadBalancer",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "ApplicationLoadBalancerTesting172A253B",
                        "LoadBalancerFullName",
                      ],
                    },
                  },
                ],
                "MetricName": "HTTPCode_Target_5XX_Count",
                "Namespace": "AWS/ApplicationELB",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m3",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "LoadBalancer",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "ApplicationLoadBalancerTesting172A253B",
                        "LoadBalancerFullName",
                      ],
                    },
                  },
                ],
                "MetricName": "RequestCount",
                "Namespace": "AWS/ApplicationELB",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;

exports[`The GuUnhealthyInstancesAlarm construct should create the correct alarm resource with minimal config 1`] = `
Object {
  "Mappings": Object {
    "testing": Object {
      "CODE": Object {
        "alarmActionsEnabled": false,
      },
      "PROD": Object {
        "alarmActionsEnabled": true,
      },
    },
  },
  "Outputs": Object {
    "ApplicationLoadBalancerTestingDnsName": Object {
      "Description": "DNS entry for ApplicationLoadBalancerTesting",
      "Value": Object {
        "Fn::GetAtt": Array [
          "ApplicationLoadBalancerTesting172A253B",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApplicationListener62E539D0": Object {
      "Properties": Object {
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "ApplicationTargetGroupTesting080C2FF2",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "ApplicationLoadBalancerTesting172A253B",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "ApplicationLoadBalancerTesting172A253B": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internal",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ApplicationLoadBalancerTestingSecurityGroup883A01A4",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          "",
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "ApplicationLoadBalancerTestingSecurityGroup883A01A4": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB TestApplicationLoadBalancerTesting8F9EA5A8",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": "test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ApplicationTargetGroupTesting080C2FF2": Object {
      "Properties": Object {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Port": 80,
        "Protocol": "HTTP",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "TargetGroupAttributes": Array [
          Object {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
        ],
        "UnhealthyThresholdCount": 2,
        "VpcId": "test",
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "UnhealthyInstancesAlarmTestingD3AE50F4": Object {
      "Properties": Object {
        "ActionsEnabled": Object {
          "Fn::FindInMap": Array [
            "testing",
            Object {
              "Ref": "Stage",
            },
            "alarmActionsEnabled",
          ],
        },
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":test-topic",
              ],
            ],
          },
        ],
        "AlarmDescription": "testing's instances have failed healthchecks several times over the last hour.
      This typically results in the AutoScaling Group cycling instances and can lead to problems with deployment,
      scaling or handling traffic spikes.

      Check testing's application logs or ssh onto an unhealthy instance in order to debug these problems.",
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Unhealthy instances for testing in ",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 6,
        "Dimensions": Array [
          Object {
            "Name": "LoadBalancer",
            "Value": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::Select": Array [
                      1,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "ApplicationListener62E539D0",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  Object {
                    "Fn::Select": Array [
                      2,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "ApplicationListener62E539D0",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  Object {
                    "Fn::Select": Array [
                      3,
                      Object {
                        "Fn::Split": Array [
                          "/",
                          Object {
                            "Ref": "ApplicationListener62E539D0",
                          },
                        ],
                      },
                    ],
                  },
                ],
              ],
            },
          },
          Object {
            "Name": "TargetGroup",
            "Value": Object {
              "Fn::GetAtt": Array [
                "ApplicationTargetGroupTesting080C2FF2",
                "TargetGroupFullName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 12,
        "MetricName": "UnHealthyHostCount",
        "Namespace": "AWS/ApplicationELB",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
